File: backend/app/schemas/geospatial.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List, Dict, Any, Union

class GeoFenceBase(BaseModel):
    name: str
    description: Optional[str] = None
    coordinates: Dict[str, Any]  # GeoJSON格式
    center: Dict[str, float]  # {lat: number, lng: number}
    radius: Optional[float] = None  # 如果是圆形围栏则需要半径
    fence_type: str  # polygon, circle

class GeoFenceCreate(GeoFenceBase):
    pass

class GeoFenceUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    coordinates: Optional[Dict[str, Any]] = None
    center: Optional[Dict[str, float]] = None
    radius: Optional[float] = None
    fence_type: Optional[str] = None

class GeoFenceInDB(GeoFenceBase):
    id: int
    owner_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        orm_mode = True

class GeoFenceResponse(GeoFenceInDB):
    pass

class GeoFenceList(BaseModel):
    items: List[GeoFenceResponse]
    total: int

class MapDataPoint(BaseModel):
    lat: float
    lng: float
    value: float
    label: Optional[str] = None

class MapBounds(BaseModel):
    north: float
    south: float
    east: float
    west: float

class MapDataResponse(BaseModel):
    points: List[MapDataPoint]
    bounds: MapBounds

class HeatmapDataResponse(MapDataResponse):
    intensity_max: float
    intensity_min: float

File: frontend/src/api/geospatial.ts
import { http } from '@/lib/http'

export interface GeoFence {
  id: number
  name: string
  description?: string
  coordinates: any // GeoJSON格式
  center: {
    lat: number
    lng: number
  }
  radius?: number
  fence_type: 'polygon' | 'circle'
  owner_id: number
  created_at: string
  updated_at: string
}

export interface MapData {
  points: Array<{
    lat: number
    lng: number
    value: number
  }>
  bounds: {
    north: number
    south: number
    east: number
    west: number
  }
}

export interface CityResponseData {
  cityData:  any;
  provinceData: any;
}

export interface CityData {
  name: string
  value: []
}

export const geospatialApi = {
  getFences: () => 
    http.get<GeoFence[]>('/api/geospatial/fences'),
  
  createFence: (data: Omit<GeoFence, 'id' | 'owner_id' | 'created_at' | 'updated_at'>) =>
    http.post<GeoFence>('/api/geospatial/fences', data),

  getMapData: () =>
    http.get<MapData>('/api/geospatial/map'),

  getHeatmapData: () =>
    http.get<MapData>('/api/v1/geospatial/heatmap'),

  getCityData: () =>
    http.get<CityResponseData>('/api/geospatial/city'),

  getCitySiteData: (cityname: string) =>
    http.get<CityData>('/api/geospatial/citysite', {
      params: { cityname }
    }),

  getCityHistoryData: (citycode: string) =>
    http.get('/api/geospatial/cityhistory', {
      params: { citycode }
    }),
}